include/drivers/edac.h

```diff
@@ -14,93 +14,80 @@
+ */
+enum edac_error_type {
+	/** Correctable error type */
+	EDAC_ERROR_TYPE_DRAM_COR = BIT(0),
+	/** Uncorrectable error type */
+	EDAC_ERROR_TYPE_DRAM_UC = BIT(1)
+};
 
 /**
  * @brief EDAC driver API
  *
  * This is the mandatory API any EDAC driver needs to expose.
  */
 __subsystem struct edac_driver_api {
-#if defined(CONFIG_EDAC_ERROR_INJECT)
 	/* Error Injection API is disabled by default */
-	edac_api_inject_set_param1_f inject_set_param1;
-	edac_api_inject_get_param1_f inject_get_param1;
-	edac_api_inject_set_param2_f inject_set_param2;
-	edac_api_inject_get_param2_f inject_get_param2;
-	edac_api_inject_get_error_type_f inject_get_error_type;
-	edac_api_inject_set_error_type_f inject_set_error_type;
-	edac_api_inject_error_trigger_f inject_error_trigger;
-#endif /* CONFIG_EDAC_ERROR_INJECT */
+	int32_t (*inject_set_param1)(const struct device *dev, uint64_t value);
+	int32_t (*inject_get_param1)(const struct device *dev, uint64_t *value);
+	int32_t (*inject_set_param2)(const struct device *dev, uint64_t value);
+	int32_t (*inject_get_param2)(const struct device *dev, uint64_t *value);
+	int32_t (*inject_set_error_type)(const struct device *dev,
+					 uint32_t value);
+	int32_t (*inject_get_error_type)(const struct device *dev,
+					 uint32_t *value);
+	int32_t (*inject_error_trigger)(const struct device *dev);
 
 	/* Error Logging  API */
-	edac_api_ecc_error_log_get_f ecc_error_log_get;
-	edac_api_ecc_error_log_clear_f ecc_error_log_clear;
-	edac_api_parity_error_log_get_f parity_error_log_get;
-	edac_api_parity_error_log_clear_f parity_error_log_clear;
+	int32_t (*ecc_error_log_get)(const struct device *dev, uint64_t *value);
+	int32_t (*ecc_error_log_clear)(const struct device *dev);
+	int32_t (*parity_error_log_get)(const struct device *dev,
+					uint64_t *value);
+	int32_t (*parity_error_log_clear)(const struct device *dev);
 
 	/* Error stats API */
-	edac_api_errors_cor_get_f errors_cor_get;
-	edac_api_errors_uc_get_f errors_uc_get;
+	int32_t (*errors_cor_get)(const struct device *dev);
+	int32_t (*errors_uc_get)(const struct device *dev);
 
 	/* Notification callback API */
-	edac_api_notify_cb_set_f notify_cb_set;
+	int32_t (*notify_cb_set)(const struct device *dev,
+			     edac_notify_callback_f cb);
 };
 
-#if defined(CONFIG_EDAC_ERROR_INJECT)
+/* Optional interfaces */
 
 /**
  * @brief Set injection parameter param1
  *
- * This parameter is used to set first error injection parameter value.
+ * Set first error injection parameter value.
  *
  * @param dev Pointer to the device structure
- * @param addr Injection address base
- * @return 0 on success, error code otherwise
+ * @param value First injection parameter
+ *
+ * @retval -ENOSYS if the optional interface is not implemented
+ * @retval 0 on success, other error code otherwise
  */
-static inline int edac_inject_set_param1(const struct device *dev,
-					 uint64_t addr)
+static inline int32_t edac_inject_set_param1(const struct device *dev,
```

Comment:
The convention for APIs that return negative POSIX error numbers is `int` and not `int32`.

Reply:
what convention exactly? this convention you are talking about conflicts with the coding guidelines adopted by Zephyr.

Reply:
I mean't a de-facto convention, i.e. an established practice that can be seen throughout the tree. Whenever you see an `int` return you can assume that it has these semantics. I just counted 66 such APIs in `include/kernel.h` alone. If we move away from `int` I don't think `int32_t` is the right choice, since it looses essential semantic information  IMO. Instead we could introduce some new `err_t` type which is known to have these semantics.

Reply:
@finikorg I think you're referencing the wrong commit. In general, I think this has project-wide implications so review comments in this PR aren't necessary the right place for the discussion. I'd consider the API or Dev meetings instead.

Reply:
@jhedberg Fixed commit.

Reply:
This has been raised before. I am of the opinion that using `int` wherever the size in bits of the variable doesn't matter is the correct approach, because it uses the native integer type of the platform. However, if this gets in the way of safety certification I agree we should consider it, and slowly transition to `int32_t` everywhere. In any case the error codes should always be documented with Doxygen so that it is clear for the user what to expect as a return value.
So to summarize my position: I'd rather use `int`, but safety certification is more important than my personal taste. If safety requires this then we need to transition to using `int32_t` everywhere for consistency, I assume then also on 64-bit platforms.

Reply:
@nashif wouldn't `size_t` be an issue as well in this case? since its size is platform-dependant, as @jhedberg observed offline.

Reply:
Btw, this also relates to #32213 , i.e. if we are going to change the commonly used error type we might want to introduce the concept of domain in the same go.

Reply:
I was looking here. Is it true that the rule has become mandatory in the 8 years from 2004 to 2012?

https://bullwhip.physio-control.com/documentation/help/reference/misra.builtin_numeric.htm

I agree with carlescufi on this point. 

Also, what do we do on 64-bit platforms with these APIs? Disallowing native types for error returns seems like a strange rule to me. You cannot stop an int32_t being promoted to a 64-bit value if that is the machine's register size. It might just add more confusion.

The page referenced above refers to variables that store some value with a meaning (speed, torque). I am not even sure that error codes fit that definition. If they do end up being stored in a data structure then using a specific length perhaps makes more sense.

Reply:
I removed that patch introducing that change `int` -> `int32_t`. I did not like it at all, just had no other option ;)